Zoom

BlueBazze opened the pull request:
## Description of the Change
Added prop `nativeZoomEnabled` which describes if the native platform should handle zoom.
Added method `setZoomRatio(zoomRatio: number)` which can be used to set the zoom ratio without the native zoom. For instance if you want to implement your own zooming functionality.

## Possible Drawbacks
Some users might get confused. With React Native, if you draw a View with Position: "Absolute" over the preview. That View would capture all of the click events. So you will need to add pointerEvents="none" to the view stretching over the preview to make touches go through that view. - Working on documentation to inform regarding this.

## Requirements
Before this can be merged, both of these PR's needs to be included in a release.
https://github.com/apivideo/api.video-ios-live-stream/pull/8
https://github.com/apivideo/api.video-android-live-stream/pull/9

## Comments
I was a little confused by the structure of the iOS side regarding props. Hopefully it works.

Also, the `setZoomRatio`, i tried implementing it as a method for now. Gotta test it to know if this is viable. Previously i have used a prop & `setNativeProps` to set this. Knowing if it works as a method is crucial because it can cause serious performance issues. Otherwise we will have to revert it to a prop.

I havent tested any of this yet. Could figure out how to add a local or github dependency on android. So decided to wait for a proper release before continuing this PR.

ThibaultBee commented on README.md:
```diff
@@ -165,6 +165,8 @@ type LiveStreamProps = {
   // Mute/unmute microphone
   isMuted: false;
+  // Enables/disables the zoom gesture handled natively
+  nativeZoomEnabled?: boolean;
```
Renames to `enablePinchedZoom` or `enableZoomOnPinche` (action first).
You could also set the property as mandatory with default value to `true`

BlueBazze replied:
Done

ThibaultBee commented on README.md:
```diff
@@ -180,6 +182,8 @@ type LiveStreamMethods = {
   // Stops the stream
   stopStreaming: () => void;
+  // Sets the zoom ratio. Can be used if you want to control the zooming.
+  setZoomRatio: (zoomRatio: number) => void;
```
Use property instead?

BlueBazze replied:
My idea was, normally with anything that is a gesture or slider. You will 99% of the time have access to a `onValueChange` Which in this case should be called continuously while the value is changing. 

Do you mean component prop? Or ref property?

As a component prop wouldnt work, because it can cause a rerender. The rerenders would pile up and eventually crash the app from what i have seen.
Ill play a little with it to see what i can do.

ThibaultBee replied:
ref property such as `isMuted`

BlueBazze replied:
What do you say if we have both? Both can be used in different cases, and the performance vary.
If you have a list of available zooms that you can switch between, it will not be a problem to pass it down as a prop because it wont be updated as often.
But if you want to achieve a smooth zoom with a slider, then the setZoomRatio will be a better choice simply because it wont rerender the react part.
Ive set up a demo of this in the example.

Updating a zoomRatio prop while sliding a slider using a state, will cause the zooming to be a little laggy, maybe only every 5-10 tick will actually be passed down to the zoom.
I would really like to see how it would perform with an animated value, but havent been able to convert that to a number.
Using the method setZoomRatio with a slider is fairly smooth. It is slightly laggy compared to the native zoom, but still barely noticeable.

What do you think?

ThibaultBee replied:
I am not an expert of RN.
I guess you have much more experiences than I do 👍 
So I am okay with implementing both ways. 
The only that triggers me is that there should be a way to get current zoom ratio (which is possible with the zoomRatio prop).

BlueBazze replied:
Alright, ill take a look at getting current zoomRatio

BlueBazze replied:
So i have now gone it over.
I am not entirely sure what you mean by getting the current zoom by the regular prop?

If i want to get the current zoomRatio from the native side. How would i do that with the prop?
Normally i would create a 
```
@ReactMethod
  fun getZoomRatio(view: ReactNativeLiveStreamView, promise: Promise) {
    try {
      val _zoomRatio: Float = view.zoomRatio
      promise.resolve(_zoomRatio)
    } catch (e: java.lang.Exception) {
      promise.reject(e)
    }
  }
```
in a native module which gets that value.
Is there a better way?

ThibaultBee replied:
Oups, sorry. I haven't answered here.

Is there a way get get the zoom value with the `zoomRatio` prop like doing thing like this:
`const zoomRatio = ref.current?.zoomRatio`
I guess no that is why we don't understand each other ;)
In this case, the `getZoomRatio` would be perfect!

If `setZoomRatio` has better performance than `zoomRatio`, could you stash pop`setZoomRatio`?

BlueBazze replied:
With setZoomRatio you would be able to do some cool stuff with RN gestures & animations. Which is not possible with the prop.

Last week i went through everything, and i couldnt find a way to do it neatly. - Wanted to get zoomRange, current zoom from the native side to js.
The way i used to do it, was to have the streamer variable in a singleton. That way you can interact with it through a module.
This way is viable. But i didnt have enough time to maintain this on top of everything else. So eventually it stopped working.
This specific version was forked from some turkish guy who deleted it from github.
https://github.com/BlueBazze/rtmp/blob/master/android/src/main/java/com/reactnativertmppublisher/RTMPModule.java

ThibaultBee replied:
Ok.
If I have understand correctly, the slider should use `setZoomRatio` instead of the `zoomRatio` props. Right?

For a getter on zoom ratio, unfortunately I don't have time to have a look on it this week. It would be nice to have but let say we drop it for now till someone asks for it.

BlueBazze replied:
Ill quickly adjust the example.

I agree. - The zoomRatio will always start at 1. And people can keep their own state when they setZoomRatio for now.

ThibaultBee commented on docs/Zoom.md:
````diff
@@ -0,0 +1,96 @@
+}
+
+export default App;
+```
````
Wahoo! Nice!

ThibaultBee commented on ios/ReactNativeLiveStreamView.swift:
```diff
@@ -153,7 +157,18 @@ class ReactNativeLiveStreamView : UIView {
     }
-    
+
+    @objc var nativeZoomEnabled: Bool {
```
Could it be a setter/getter here like `isMuted`?

ThibaultBee commented:
Hi,
I have released both Android and iOS live stream library.

I would rather see a `zoomRatio` property (with getter and setter) than a `setZoomRatio` method.

BlueBazze commented:
Quick note. Noticed that zoom still stopped the preview on sdk 30. But it does work as expected on Android 9.
https://drive.google.com/file/d/1kf2ckaWel-aH_QCuCEp9GhpEWGwBooMQ/view?usp=sharing

ThibaultBee commented:
Hum... I don't really know... Does your phone does support the zoom?
Could you try directly with StreamPack?

BlueBazze commented:
Ive set up a "demo" using a zoom slider. I will clean all this up again afterwards.

Try it out on android. havent changed iOS to accept the zoomRatio prop yet.
Using a state makes it quite laggy compared to using the ref method.
Also tried to use an animated value. But that really didnt wanted to be converted into a number. Probably could change this in the native code to accept Animated.Value.
https://github.com/BlueBazze/api.video-reactnative-live-stream/blob/main/example/src/App.tsx#L227-L228

BlueBazze commented:
> Hum... I don't really know... Does your phone does support the zoom? Could you try directly with StreamPack?

Same issue
https://drive.google.com/file/d/1jlraBfT5nW58z-KurxRbWWgECB2Cr-RE/view?usp=sharing

ThibaultBee commented:
> > Hum... I don't really know... Does your phone does support the zoom? Could you try directly with StreamPack?
> 
> Same issue. Dont know if this is related to Nokia, or if its generally a sdk 30 problem. As previously said. I tested on a Samsung A22 running Android 11 which was working. https://drive.google.com/file/d/1jlraBfT5nW58z-KurxRbWWgECB2Cr-RE/view?usp=sharing

I tested with a very old Nexus 5 but everything was ok.
What if you put breakpoint in `zoomRatio` or in the `availableZoomRange`?

BlueBazze commented:
> > > Hum... I don't really know... Does your phone does support the zoom? Could you try directly with StreamPack?
> > 
> > 
> > Same issue. Dont know if this is related to Nokia, or if its generally a sdk 30 problem. As previously said. I tested on a Samsung A22 running Android 11 which was working. https://drive.google.com/file/d/1jlraBfT5nW58z-KurxRbWWgECB2Cr-RE/view?usp=sharing
> 
> I tested with a very old Nexus 5 but everything was ok. What if you put breakpoint in `zoomRatio` or in the `availableZoomRange`?
> 
> I have the feeling that somehow it does not support any zoom but we set a zoom of 1.0f that makes the preview crashes.

The zooming is working in the default camera app.
Followed the stack. The preview stops when it returns to this line 
https://github.com/ThibaultBee/StreamPack/blob/fa05dcc939ed317919c2a297e9d95dea6b5cf2f0/core/src/main/java/io/github/thibaultbee/streampack/internal/sources/camera/CameraExecutorManager.kt#L71

ThibaultBee commented:
What are the values of `zoomRatio` and `availableZoomRange`?

BlueBazze commented:
When i debugged it, the value were 4x, and i think it was the max (not sure, and cant debug this again before monday).
The max zoom in the default camera app were 4x for sure.

BlueBazze commented:
Logging the zoomRatioRange gives: 1.0 - 4.0 (lower - upper)
Zooming in slowly - At a zoomratio of 1.02, the preview does not stop. At 1.17 the preview havent stopped. But around 1.3-1.5 the preview stops.
I havent tried to see if zooming works depending on which camera facing is active.

But, if we can figure out if this is a Nokia problem or a general android problem. It might just be an issue isolated to this model.
Do you have any Nokia devices at api.video? Or any employees using such a device?
If this is isolated to this model. Then it could be that it wont be a problem, it is a 720p camera, probably not many will want to stream with this device. But if it is related to Nokia's distro or even specifically their android 11.

ThibaultBee commented:
Wahoo
I have already look on google for Nokia camera2 API and zoom but couldn't find anything. Unfortunately we don't have any Nokia phone at api.video. 
Unfortunately, we can't say if it comes from Nokia distro, Android 11, StreamPack or your device...

BlueBazze commented:
> Wahoo I have already look on google for Nokia camera2 API and zoom but couldn't find anything. Unfortunately we don't have any Nokia phone at api.video. Unfortunately, we can't say if it comes from Nokia distro, Android 11, StreamPack or your device...

So probably before making any changes. We can create an issue on StreamPack with what we know at this point. If other people should come across this issue we can get to know more about it.

ThibaultBee commented on example/src/App.tsx:
```diff
@@ -52,6 +55,10 @@ export default function App() {
+  if (ref.current) {
+    console.log(ref.current);
+  }
+
```
Debug trace: to remove

BlueBazze commented on android/src/main/java/video/api/reactnative/livestream/ReactNativeLiveStreamViewManager.kt:
```diff
@@ -73,4 +77,14 @@ class ReactNativeLiveStreamViewManager : SimpleViewManager<ReactNativeLiveStream
+  }
+
+  @ReactProp(name = ViewProps.ZOOM_RATIO)
+  fun enablePinchedZoom(view: ReactNativeLiveStreamView, zoomRatio: Double) {
```
Shh

BlueBazze commented:
- Removed the `setZoomRatio` command 
Maybe i misunderstood the last review comment on the method in the readme file
Stashed it so it can be reapplied quickly.

- Moved the native component into its seperate file - Got tired of getting the component registred twice
In development mode, reloading a file containing requireNativeComponent will cause RN to register it again. Meaning it has been registered twice. This is easily reproducible with just hot reloading the file.
Also good practice, see 
https://reactnative.dev/docs/new-architecture-library-intro#move-the-call-to-requirenativecomponent-to-a-separate-file
https://reactnative.dev/docs/native-modules-android#better-native-module-export
https://reactnative.dev/docs/native-components-android#6-implement-the-javascript-module

- Had problems with enabling/disabling the zoom gesture on iOS
Found a property on `zoomGesture` called `isEnabled` which made everything a lot simplier. It does mean that during init we add the gesture to the preview. And then disable or enable the gesture itself when setting `enablePinchedZoom`.

BlueBazze commented:
Got an error when building the iOS app. XCode wanted me to unwrap the newValue of audio & videoConfig.
https://github.com/apivideo/api.video-reactnative-live-stream/blob/main/ios/ReactNativeLiveStreamView.swift#L101

Adding `!` stops the build error. But havent commited it since i dont know enough about this.
```
liveStream.audioConfig = newValue!
```

ThibaultBee commented:
Hi,

I just read/built/test your project. Everything is perfect. 


> Got an error when building the iOS app. XCode wanted me to unwrap the newValue of audio & videoConfig. https://github.com/apivideo/api.video-reactnative-live-stream/blob/main/ios/ReactNativeLiveStreamView.swift#L101
> 
> Adding `!` stops the build error. But havent commited it since i dont know enough about this.
> 
> ```
> liveStream.audioConfig = newValue!
> ```

Ok, I fixed it. `newValue` was not supposed to be optional.

> * Removed the `setZoomRatio` command
>       Maybe i misunderstood the last review comment on the method in the readme file
>       Stashed it so it can be reapplied quickly.

=> See answer in the README.md review thread.

>     * Moved the native component into its seperate file - Got tired of getting the component registred twice
>       In development mode, reloading a file containing requireNativeComponent will cause RN to register it again. Meaning it has been registered twice. This is easily reproducible with just hot reloading the file.
>       Also good practice, see
>       https://reactnative.dev/docs/new-architecture-library-intro#move-the-call-to-requirenativecomponent-to-a-separate-file
>       https://reactnative.dev/docs/native-modules-android#better-native-module-export
>       https://reactnative.dev/docs/native-components-android#6-implement-the-javascript-module

Ok, as I said I am not familiar with RN. Obviouslym I am okay with this. Thank you for your help 👍 

>     * Had problems with enabling/disabling the zoom gesture on iOS
>       Found a property on `zoomGesture` called `isEnabled` which made everything a lot simplier. It does mean that during init we add the gesture to the preview. And then disable or enable the gesture itself when setting `enablePinchedZoom`.

Ok, I read that in the code. I also tested it. Work perfectly :)

ThibaultBee commented:
Ok, I am done with the review :)
Ready?

Great job! Thank you very much for your help 🥇

BlueBazze commented:
Yeah, yeet it
